### ISSUE TITLE
Configuration Question

### ISSUE THREAD

**Issue by mkumpf** [Mcpoowl/SFTPModule#6]
Paul,

I am struggling with getting this working - most of the issue is that I am an FTP newbie.

I installed the module, configured it and got it working with no issues on my local machine.  I pointed it to an ftp site I set up at www.brickftp.com and was able to retrieve files with no issues.

When I deploy to the cloud, I get errors in the log.  error messages look something like:

Apr 21 14:54:09.377 127.0.0.1 tr10000: ERROR - (S)FTP: (1/63) org.apache.commons.net.MalformedServerReplyException: Could not parse response code.
Apr 21 14:54:09.379 127.0.0.1 tr10000: ERROR - (S)FTP: (2/63) Server Reply: Protocol mismatch.
Apr 21 14:54:09.379 127.0.0.1 tr10000: ERROR - (S)FTP: (3/63) 
Apr 21 14:54:09.379 127.0.0.1 tr10000: ERROR - (S)FTP: (4/63)   at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:337)
Apr 21 14:54:09.379 127.0.0.1 tr10000: ERROR - (S)FTP: (5/63)   at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:294)
Apr 21 14:54:09.379 127.0.0.1 tr10000: ERROR - (S)FTP: (6/63)   at org.apache.commons.net.ftp.FTP.sendCommand(FTP.java:483)
Apr 21 14:54:09.380 127.0.0.1 tr10000: ERROR - (S)FTP: (7/63)   at org.apache.commons.net.ftp.FTP.sendCommand(FTP.java:608)
Apr 21 14:54:09.380 127.0.0.1 tr10000: ERROR - (S)FTP: (8/63)   at org.apache.commons.net.ftp.FTP.sendCommand(FTP.java:582)
Apr 21 14:54:09.380 127.0.0.1 tr10000: ERROR - (S)FTP: (9/63)   at org.apache.commons.net.ftp.FTP.quit(FTP.java:864)

etc. etc.  I attached the whole error stack to this issue as a file.
[FTPErrorMessage.txt](https://github.com/Mcpoowl/SFTPModule/files/230632/FTPErrorMessage.txt)

I have not set up SSH keys or any other security settings in the cloud installation.  Can you give me a beginner's run down of how to go about doing that?  Also, if you see something in the errors that may give me a direction to go in to resolve this, I would really appreciate it.

Thanks,

Mike

**Comment by Mcpoowl** [comment]
Hi Mike,

If you've used the same settings as locally, my guess is that is something
being blocked by the cloud. I'll try to test it a bit tomorrow. I'll keep
you posted!
Op 21 apr. 2016 22:04 schreef "mkumpf" notifications@github.com:

**Comment by mkumpf** [comment]
Paul

Thanks for getting back to me - your help is much appreciated!

Mike

**Comment by mkumpf** [comment]
BTW, in case you want to use the same FTP site I am using:

brook.brickftp.com

username:  spstest

password:  password

directory is called out (there are 2 documents in that directory)

**Comment by RCatersels** [comment]
Could it be that without the SSL it will fall back to normal FTP and thus not function in the cloud?

**Comment by RCatersels** [comment]
Mike, I always use WinSCP to test the setup. They have some good tutorials on how to create keys and do a test setup. See here: [https://winscp.net/eng/docs/public_key] . To be able to use the keys you have to make them openssh compatible.  In puttygen you can do that by clicking conversion / export to openssh. Now place the private key in the resources/SFTP map if your Mendix project directory. Now change the know_hosts file and add the brook.brickftp.com site in that file. Each SFTP server needs a new line. 
And some last double checking. SFTP is something different then FTPS. Make sure that you connect with an SFTP server. 

Ronald

**Comment by Mcpoowl** [comment]
Mike,

Can you (preferably with screenshots) tell me what you're settings are and if you can see your documents in the application

**Comment by Mcpoowl** [comment]
Mike,

I managed to get it working in a Sandbox environment, without really doing anything different as to my local setup.

Here is my settings screen using your FTP credentials

![image](https://cloud.githubusercontent.com/assets/2244753/14734599/d82d00da-086b-11e6-955e-328346895505.png)

And the actual document overview

![image](https://cloud.githubusercontent.com/assets/2244753/14734607/e4dca146-086b-11e6-9534-13b620a0f867.png)

Let me know if this works for you. Please bear in mind that this is the regular FTP connection, so no SFTP. For that you need to have a private.key and a known_hosts file in your resources folder which match with your server.

**Comment by mkumpf** [comment]
Paul and Ronald

Thanks for the assistance!

@Mcpoowl - what Mendix version did you use for this testing
@RCatersels  - I will work with the ssh key resources you included in your comments.  I had a key in the resources folder, but not sure if it was created correctly.

I'll keep you posted on my progress with this.

Miek

**Comment by Mcpoowl** [comment]
Hi Mike,

I tried a late Mx5 version since the module isn't Mendix 6 compatible from
the Appstore. I used 5.21.1 if I recall correctly.

Good luck!
Paul
Op 22 apr. 2016 15:22 schreef "mkumpf" notifications@github.com:

Paul and Ronald

Thanks for the assistance!

@Mcpoowl https://github.com/Mcpoowl - what Mendix version did you use for
this testing
@RCatersels https://github.com/RCatersels - I will work with the ssh key
resources you included in your comments. I had a key in the resources
folder, but not sure if it was created correctly.

I'll keep you posted on my progress with this.

Miek

**Comment by mkumpf** [comment]
@RCatersels & @Mcpoowl  I generated a key using PuttyGEN.  Then I converted it to an OpenSSH key - which generated a file.  I assume this file is my private key - is this correct?  I also saved the public key and private key on my machine.  The Resources directory for my project does not have a folder for SFTP - so I am going to create one, create a text file in that directory called known_hosts (does this file need an extension) and then put one line in that file with the following text:  brook.brickftp.com  Finally, I will past the generated OpenSSH key into this resources file.  What do I do with the public key?  I searched out some known_hosts file examples on line and they seemed to have the domain followed by a key - is that the public key?  Do I need to paste my public key into the known_hosts file?

Anything else I am missing?

Thanks again for your help!

**Comment by RCatersels** [comment]
Strange that you do not have a SFTP directory in your resource folder. The private key is for the Mendix server, the public key you hand out to the other side. That key must be added to the SFTP server where you connect with. The known_hosts has no extension and only holds the server names you want to connect with. That can be an IP address or the name of the server. There is nothing else you need to put there. If you still not get it to work I will see if I can make you a demo project and database so you can check out the setup.
Have a nice weekend,
Ronald

**Comment by mkumpf** [comment]
Ronald
I have tried a number of different configurations and setups without success.  I created public and private keys and think they are in the correct places.  If you have time to set up a small demo project, that would be great!  Let me know if you need any additional info from me.
Mike

**Comment by Mcpoowl** [comment]
Hi Mike,

Quick question. Do you want to connect via SFTP, or regular FTP.
Ronald (thanks for all the responses btw!) is talking about settings regarding an SFTP connection, however, when connecting through regular FTP these files aren't neccesary. As said the above screenshots that I posted are done through FTP.

Can you verify that works for you? Or do you want to connect with SFTP? I'm asking since the error you posted contains a part of the stacktrace of apache.commons, which is the FTP library, and has "nothing" to do with SFTP (which uses the jsch library)

Just to make things clear :)

**Comment by mkumpf** [comment]
Paul

Actually, I need to do both.  For this app, I have 1 vendor that I can
connect with via FTP and another that requires connection via SFTP.  I
haven't gotten either to work in the acceptance instance of my Mendix cloud
deployment.  I have gotten FTP to via the modeler to work but not SFTP.
When I added the certificate and known hosts to the modeler and tried to
connect via SFTP, the modeler never responds, no error message, just no
response when I call the Java action.

Let me know if I can provide any additional info.

Mike

**Comment by Mcpoowl** [comment]
Can you please verify if the above setup works for you in your cloud environment using FTP to connect. Since I got it working in the sandbox It should work in a licensed cloud node aswell..

As far as SFTP goes, it's clear to me that @RCatersels has a lot more knowledge in that area compared to me :-)

**Comment by mkumpf** [comment]
Paul & Ronald,

I wanted to do a bit more extensive testing before coming back to this thread.  I tested quite a bit using brook.brickftp.com.  Then I decided to use another test ftp site called test.rebex.net.  You can see the test specs at that site.  I compiled my results in a spreadsheet.  I should also note that I patched the FTP module according to this forum thread:  https://mxforum.mendix.com/questions/13602/FTPS-FTP-over-SSL-possible  thinking that this might provide a different result.  You'll see the outcome of the testing in the file I am attaching here:  
[FTP Results.xlsx](https://github.com/Mcpoowl/SFTPModule/files/235327/FTP.Results.xlsx)

Finally, I looked at support.mendixcloud.com and I noticed this template for a new ticket:

![supportscreencapture](https://cloud.githubusercontent.com/assets/3680543/14797372/7973627c-0b00-11e6-84a9-daac9b4138dc.PNG)

Those are my datapoints from todays work.  Any help you can provide would be greatly appreciated.

**Comment by RCatersels** [comment]
Mike,
Note that in the Mendix cloud only SFTP will ever work. All the other versions is not allowed by Mendix because all the data needs to go over a single port. All the other protocols send data over multiple ports. So the other protocols only work in on premis sollutions where the security is handled by the customer. See also this site [http://www.jscape.com/blog/bid/75602/Understanding-Key-Differences-Between-FTP-FTPS-and-SFTP] and especially the firewall parts of that text. Then you know why Mendix does not allow the other protocols.
Regards,
Ronald

**Comment by mkumpf** [comment]
Ron,
Yes, I am aware of SFTP being the only option in the Mendix cloud.  I just did the testing I did in order to be thorough - sometimes I get results that are different than what I expect when I test everything.  And often I will learn something new.
Mike

**Comment by mkumpf** [comment]
Paul and Ronald,

Another datapoint that is a bit of a puzzle.  For testing purposes, I created a test app with only the (S)FTP module in it.  This worked for both FTP and SFTP on my local machine.  I then deployed this to a sandbox, and in the sandbox it also works for both FTP and SFTP.  I put the same private.key and known hosts file in this test app as I used in my regular app. You can test it out here:  https://testftpapp.mxapps.io/index.html

A couple of questions:
- any ideas about how to identify environment differences between sandbox and mendix cloud that would account for the different results?  If I need to file a ticket with support, I am not sure what to ask them to change.  Or, perhaps, there are changes I can make to my Mendix cloud environments
- Interesting that FTP and SFTP both worked in the sandbox.  I assumed that Sandbox apps are deployed in the standard Mendix cloud, but maybe there are some differences between sandbox and Mendix cloud that enable FTP to work.

If you have any thoughts on these questions or suggested next steps, I would appreciate any ideas.

P.S. Paul, I know you tested a similar test app previously.  I wasn't ignoring your work, just trying to narrow down what the issue may be, so thought I would test it out myself.

**Comment by RCatersels** [comment]
Mike, is it a teamserver project? Could you add me to it so I can look at the app. And do you have credentials for logging into the app? Or did you add our Mendix profile?

**Comment by mkumpf** [comment]
Ronald - I don't seem to have your email address to invite you to the teamserver project....

**Comment by mkumpf** [comment]
So, this will be a bit of a rambling post.

After creating the test project, I decided to put the USE_ME links into the app I am working on and see if they worked there.  They did - both FTP and SFTP worked in the cloud as well as locally.  Then I decided to try the FTP capabilities I had build into the app that did not work previously in the cloud....and they worked.  Both FTP and SFTP.  While I am happy that they worked, this is not a great outcome.  After all, if something works today that didn't work yesterday, it could just as easily stop working tomorrow!  So here is what I have learned:
- both FTP and SFTP appear to work in the cloud in my app and in the test app I created
- I made some modifications to the SFTP java actions in my app, but in the test app, I used the out of the box module.  So it doesn't seem as if my modifications are responsible for it working or not working.
- I have tested these against test.rebex.net.  Will now test against some other ftp servers to see if they work for those also

Not sure any issues have really been resolved....I don't see any cloud release notes that may have had a bearing on this.

**Comment by RCatersels** [comment]
Hello Mike, 

Sorry for the delay, but we had kingsday yesterday and thus some free time. I played with the out of the box version a bit. There is one thing in the current module I am not so happy about and was according to my memory different in the last time I worked with the module. When selecting SFTP and you get the option for the PrivateKeyFileName you should also be able to enter the passphrase. Because making a privat / public key pair without a passphrase is plain stupid and a security risk. The option to enter a passphrase is only possible when selecting the show user credentials. 

In my opinion the user credential part is only needed when you want to setup keypairs for individual users and not necesarry when seting up a global config. 
The other part is that error handling should be better. In the front I could get the message that no files are found while the real error was a USERAUTH fail. 

I hope Paul can chime in here. I think we should create a Mx6 version of the module that solves some of these issues.

Regards,

Ronald

**Comment by RCatersels** [comment]
@Paul, what is the current status of the module regarding intelectual property rights. I seem to recall you switched from employer. Is this still your module or are we in muddy water here?

**Comment by Mcpoowl** [comment]
Hi Mike, Ronald,

You're correct. I did switch employers and that does leave us in some muddy water. I've spoken with my previous employer about maintaining and updating the module to Mendix 6, but after some emailing we hit a dead end.

Basically there were a few options:
They would use the source and update the module and update it in the appstore, causing it to be listed as their appstore project (which I am fine with)
I update the module (with your help of course :-) ) and upload it as a new version to the appstore, causing 3 different FTP modules to be present in the appstore: The MX4 from CapGemini, the Mx5 from Pink and the Mx6 as the new module.

So your muddy water comment hit the spot, I can try discussing it with them again, hoping for more response. 

Regarding your other comments:
- Passphrase: That shouldn't be the case. At least not intentional. The passphrase option should always be available when selecting SFTP. If that's not correct, then that's a bug for sure!
- Error Handling: I'm totally with you on that, that was one of the items on my list if I was to update the module.
- The User credential option was introduced out of personal gain, where a couple of people I know used an FTP and they'd have different logins. The boolean provides this functionality. I'd figure it made a nice addition.

**Comment by RCatersels** [comment]
Ok, then lets make a new Mx6 version on Github and do a pull request. Then we will find out if they publish the new version in the appstore. Otherwise people still can grab the Mx6 version  straight from github. Next week I will be on vacation, but I will start when I get back.

**Comment by RCatersels** [comment]
And I think I will raise this issue in one of the MVP groups. I know that Erwin 't Hoen had the same problem when he switched from employee. Allthough I have not a clear sollution on this. It should be possible that if the main branch is not maintained by the 'owner' a branch should become the main and update rights should be transferred to a new person. Or may be the best sollution is to just create a new app in the appstore. This might give the problem with multiple apps with the same functionality but then the user of the app can decide which one to use.

**Comment by Mcpoowl** [comment]
This repository is the original one, and mine, so we can use this one and take matters from there. That being said, Pink can't use this repository to upload a new version to the appstore, because they don't have access to my Github account. But we'll cross that bridge when we get there.

I'll talk to Pink to see if they want to co-develop something
